### DISCUSSION TITLE
Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set.

### DISCUSSION THREAD

**Discussion by MohsinUddinAbir** [prisma/prisma#18958] [Q&A]
### Bug description

When I try to use the local MongoDB server on Windows it gives me this error: Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set.  Also the link: https://pris.ly/d/mongodb-replica-set does not give me any solution.

Can you develop an option for development with MongoDB using a local server on Windows like Mongoose, it is much more annoying to set up a replica, and we also don't like to use MongoDB Atlas, It also consumes more time.

### Environment & setup

- OS: Windows
- Database:  MongoDB
- Node.js version: 16.16.0

### Prisma Version

```
4.12.0
```

**Comment by ludralph** [comment]
Hi @MohsinUddinAbir 👋 

Can you please take a look at this [closed issue](https://github.com/prisma/prisma/issues/8266)? There are workarounds in there that could help you set up a replica set locally.

 You can also follow the [Mongodb link](https://www.mongodb.com/docs/manual/tutorial/convert-standalone-to-replica-set/) to set up a replica set from a standalone MongoDB server

**Reply by ludralph** [reply]
Hi @MohsinUddinAbir 

Do you have any more questions about this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏

**Comment by kumail-talverse** [comment]
the same issue encounter 
![image](https://github.com/prisma/prisma/assets/125370937/6d8a9ab5-e283-454d-affd-1ad80af38a27)

**Comment by aprilmintacpineda** [comment]
If you're still encountering this:

## Install prerequisites

1. Install docker
2. Run docker

## Create local replica set

### Create docker network for replicate set

`docker network create mongoCluster`

### Create replica set

1.`docker run -d -p 27017:27017 --name mongo1 --network mongoCluster mongo:latest mongod --replSet myReplicaSetName --bind_ip localhost,mongo1`
2. `docker run -d -p 27018:27017 --name mongo2 --network mongoCluster mongo:latest mongod --replSet myReplicaSetName --bind_ip localhost,mongo2`
3. `docker run -d -p 27019:27017 --name mongo3 --network mongoCluster mongo:latest mongod --replSet myReplicaSetName --bind_ip localhost,mongo3`

### Initiate replicate set

```
docker exec -it mongo1 mongosh --eval "rs.initiate({
 _id: \"myReplicaSetName\",
 members: [
   {_id: 0, host: \"mongo1\"},
   {_id: 1, host: \"mongo2\"},
   {_id: 2, host: \"mongo3\"}
 ]
})"
```

### Update /etc/host to include links for mong1, mongo2, mongo3

Edit `/etc/host` and append the following:

```
127.0.0.1 mongo1
127.0.0.1 mongo2
127.0.0.1 mongo3
```

### Check status

You can run `docker exec -it mongo1 mongosh --eval "rs.status()"` and it should return something like the following:

<summary>
<details>

```
Current Mongosh Log ID:	6528adb69f801781a9bc0d09
Connecting to:		mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+2.0.1
Using MongoDB:		7.0.2
Using Mongosh:		2.0.1

For mongosh info see: https://docs.mongodb.com/mongodb-shell/

------
   The server generated these startup warnings when booting
   2023-10-13T02:35:49.994+00:00: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine. See http://dochub.mongodb.org/core/prodnotes-filesystem
   2023-10-13T02:35:50.773+00:00: Access control is not enabled for the database. Read and write access to data and configuration is unrestricted
   2023-10-13T02:35:50.773+00:00: vm.max_map_count is too low
------

{
  set: 'myReplicaSetName',
  date: ISODate("2023-10-13T02:38:46.688Z"),
  myState: 1,
  term: Long("1"),
  syncSourceHost: '',
  syncSourceId: -1,
  heartbeatIntervalMillis: Long("2000"),
  majorityVoteCount: 2,
  writeMajorityCount: 2,
  votingMembersCount: 3,
  writableVotingMembersCount: 3,
  optimes: {
    lastCommittedOpTime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
    lastCommittedWallTime: ISODate("2023-10-13T02:38:40.541Z"),
    readConcernMajorityOpTime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
    appliedOpTime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
    durableOpTime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
    lastAppliedWallTime: ISODate("2023-10-13T02:38:40.541Z"),
    lastDurableWallTime: ISODate("2023-10-13T02:38:40.541Z")
  },
  lastStableRecoveryTimestamp: Timestamp({ t: 1697164690, i: 1 }),
  electionCandidateMetrics: {
    lastElectionReason: 'electionTimeout',
    lastElectionDate: ISODate("2023-10-13T02:36:30.454Z"),
    electionTerm: Long("1"),
    lastCommittedOpTimeAtElection: { ts: Timestamp({ t: 1697164580, i: 1 }), t: Long("-1") },
    lastSeenOpTimeAtElection: { ts: Timestamp({ t: 1697164580, i: 1 }), t: Long("-1") },
    numVotesNeeded: 2,
    priorityAtElection: 1,
    electionTimeoutMillis: Long("10000"),
    numCatchUpOps: Long("0"),
    newTermStartDate: ISODate("2023-10-13T02:36:30.508Z"),
    wMajorityWriteAvailabilityDate: ISODate("2023-10-13T02:36:31.065Z")
  },
  members: [
    {
      _id: 0,
      name: 'mongo1:27017',
      health: 1,
      state: 1,
      stateStr: 'PRIMARY',
      uptime: 177,
      optime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
      optimeDate: ISODate("2023-10-13T02:38:40.000Z"),
      lastAppliedWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      lastDurableWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      syncSourceHost: '',
      syncSourceId: -1,
      infoMessage: '',
      electionTime: Timestamp({ t: 1697164590, i: 1 }),
      electionDate: ISODate("2023-10-13T02:36:30.000Z"),
      configVersion: 1,
      configTerm: 1,
      self: true,
      lastHeartbeatMessage: ''
    },
    {
      _id: 1,
      name: 'mongo2:27017',
      health: 1,
      state: 2,
      stateStr: 'SECONDARY',
      uptime: 146,
      optime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
      optimeDurable: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
      optimeDate: ISODate("2023-10-13T02:38:40.000Z"),
      optimeDurableDate: ISODate("2023-10-13T02:38:40.000Z"),
      lastAppliedWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      lastDurableWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      lastHeartbeat: ISODate("2023-10-13T02:38:46.609Z"),
      lastHeartbeatRecv: ISODate("2023-10-13T02:38:45.564Z"),
      pingMs: Long("0"),
      lastHeartbeatMessage: '',
      syncSourceHost: 'mongo1:27017',
      syncSourceId: 0,
      infoMessage: '',
      configVersion: 1,
      configTerm: 1
    },
    {
      _id: 2,
      name: 'mongo3:27017',
      health: 1,
      state: 2,
      stateStr: 'SECONDARY',
      uptime: 146,
      optime: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
      optimeDurable: { ts: Timestamp({ t: 1697164720, i: 1 }), t: Long("1") },
      optimeDate: ISODate("2023-10-13T02:38:40.000Z"),
      optimeDurableDate: ISODate("2023-10-13T02:38:40.000Z"),
      lastAppliedWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      lastDurableWallTime: ISODate("2023-10-13T02:38:40.541Z"),
      lastHeartbeat: ISODate("2023-10-13T02:38:46.595Z"),
      lastHeartbeatRecv: ISODate("2023-10-13T02:38:45.568Z"),
      pingMs: Long("0"),
      lastHeartbeatMessage: '',
      syncSourceHost: 'mongo1:27017',
      syncSourceId: 0,
      infoMessage: '',
      configVersion: 1,
      configTerm: 1
    }
  ],
  ok: 1,
  '$clusterTime': {
    clusterTime: Timestamp({ t: 1697164720, i: 1 }),
    signature: {
      hash: Binary.createFromBase64("AAAAAAAAAAAAAAAAAAAAAAAAAAA=", 0),
      keyId: Long("0")
    }
  },
  operationTime: Timestamp({ t: 1697164720, i: 1 })
}
```

</details>
</summary>

### Connect

Connect through one of the following:

```
mongodb://localhost:27017,localhost:27018,localhost:27019/?replicaSet=myReplicaSetName
mongodb://localhost:27017,localhost:27018,localhost:27019/my-database-name?replicaSet=myReplicaSetName
```

**Reply by seAfnan** [reply]
Any solution without docker?
I am using mongodb locally in nestjs project and getting this error.

````
Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set. https://pris.ly/d/mongodb-replica-set
PrismaClientKnownRequestError:
Invalid `this.prisma.user.create()` invocation in
...\src\user\user.service.ts:21:44

**Reply by GuusDb** [reply]
Hey thank you for your solution everything worked fine until the connection string I get a 

 message: 'Kind: Server selection timeout: No available servers. Topology: { Type: ReplicaSetNoPrimary, Set Name: mongoCluster }, labels: {}'
 
 error whilst the previous
 docker exec -it mongo1 mongosh --eval "rs.status()" command worked as intended.
 
 Could you say where it went wrong? thanks!

**Comment by chk7964** [comment]
**Also i am facing same issue-** Prisma needs to perform transactions, which requires your MongoDB server to be run as a replica set. https://pris.ly/d/mongodb-replica-set

**Comment by edw19** [comment]
Hello Im having the same problem, its necesary use docker for resolve this issue ? Its does not exist a better solution ?

**Comment by rahul05sharma** [comment]
I am facing this same issue,
We need a proper fix. Not everyone has docker setup.
Someone please reopen this issue!

**Comment by franroav** [comment]
Simply add ?authSource=admin&retryWrites=true&w=majority to the end of your DATABASE_URL in .env file.

**Comment by ni-null** [comment]
I also encountered a problem
but not use Docker. I will record the solution here

## Environment & setup

OS: Windows 11
Database: MongoDB 8.0 install use msi
Node.js version: v20.12.2
Prisma: 5.20.0
Test on : next.js 14.2.14


### mongod.cfg 
use  administrato  to edit  file ，change MongoDB version path for yourself 
```
C:\Program Files\MongoDB\Server\8.0\bin\mongod.cfg 
```

```
replication:
  replSetName: rs0
```

run the command as an administrator，restart  MongoDB 
```
net stop MongoDB
net start MongoDB
```


### mongosh 

download mongosh  
https://www.mongodb.com/try/download/shell

cd to mongosh.exe  path 
use cmd 
```
mongosh.exe use
rs.initiate()
rs.status()
```
![Snipaste_2024-10-15_12-45-17](https://github.com/user-attachments/assets/ef47f089-6b59-40a1-97ec-bb932ef97a6e)




### try 

```
const newUser = await prisma.yourModel.create({
    data: {
      name: "John Doe",
      email: "john@example.com",
    },
  })

  const users = await prisma.yourModel.findMany()

  console.log(users)

```

i use MongoDB Compass check data 
![Snipaste_2024-10-15_12-47-07](https://github.com/user-attachments/assets/c5cfd0f1-e035-4b87-a999-40ffd6a7a5ef)

my schema.prisma 

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
    url      = "mongodb://localhost:27017/next_db"
}

model YourModel {
  id    String @id @default(auto()) @map("_id") @db.ObjectId
  name  String
  email String @unique
}


```

**Reply by qaisarhabib** [reply]
woking perfect for me thanks man

**Reply by bamiendev** [reply]
You are my savior! I wonder if there's a similar way to fix it when deploying on Ubuntu?

**Comment by masterbater** [comment]
https://www.mongodb.com/docs/atlas/cli/current/atlas-cli-deploy-docker/

**Comment by hosmanoglu** [comment]
i just use for stupid project
 do i really need replica for instert some basic documents?

**Reply by ludralph** [reply]
Hi @hosmanoglu 

For MongoDB with Prisma, you do need a replica set, even for basic operations like inserting documents. This is because Prisma uses transactions internally to support nested writes, and MongoDB requires a replica set for transactions.

- https://www.prisma.io/docs/orm/overview/databases/mongodb#example

**Reply by hosmanoglu** [reply]
hi man thanks for answering

i wonder what if we have simple future flag 
with flag we just perform simple things like insert read etc.
i mean what if prisma just translate prisma code js code to  mongo own language and run it

**Reply by ludralph** [reply]
Hi @hosmanoglu 

You can open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for our engineering team to evaluate. Please ensure you document your usecase.

**Reply by masterbater** [reply]
> hi man thanks for answering
> 
> i wonder what if we have simple future flag 
> with flag we just perform simple things like insert read etc.
> i mean what if prisma just translate prisma code js code to  mongo own language and run it

You can actually create a replica set with just one mongodb server, its a matter of config, or to ease setting up use docker images that can be configured into replica
